File: frontend/lib/show/performanceVoices.ts
import type { PerformanceSpeaker } from "@/lib/types/director";
import type { ScriptSpeaker } from "@/lib/types/script";

const DEFAULT_POOL: ScriptSpeaker[] = ["AI_A", "AI_B", "narrator"];

function normalizePool(pool?: PerformanceSpeaker[]): ScriptSpeaker[] {
  if (!pool?.length) return DEFAULT_POOL;
  const valid = pool.filter((speaker): speaker is ScriptSpeaker =>
    DEFAULT_POOL.includes(speaker as ScriptSpeaker)
  );
  return valid.length > 0 ? valid : DEFAULT_POOL;
}

/** Assigns Stimme A / B / Erzähler per sentence — never dramaturg speakers. */
export function speakerForPerformanceSentence(
  beatSpeaker: ScriptSpeaker,
  sentenceIndex: number,
  beatOrder: number,
  pool?: PerformanceSpeaker[]
): ScriptSpeaker {
  const speakers = normalizePool(pool);
  const base = speakers.indexOf(beatSpeaker);
  const start = base >= 0 ? base : beatOrder % speakers.length;
  return speakers[(start + sentenceIndex) % speakers.length];
}

File: backend/app/services/tts/performance_voices.py
from app.schemas.script import ScriptSpeaker

PERFORMANCE_SPEAKERS: tuple[ScriptSpeaker, ...] = ("AI_A", "AI_B", "narrator")


def performance_speaker_for_sentence(
    beat_speaker: ScriptSpeaker | str,
    sentence_index: int,
    beat_order: int,
    pool: list[ScriptSpeaker] | None = None,
) -> ScriptSpeaker:
    speakers: tuple[ScriptSpeaker, ...]
    if pool:
        valid = [s for s in pool if s in PERFORMANCE_SPEAKERS]
        speakers = tuple(valid) if valid else PERFORMANCE_SPEAKERS
    else:
        speakers = PERFORMANCE_SPEAKERS

    try:
        start = speakers.index(beat_speaker)  # type: ignore[arg-type]
    except ValueError:
        start = beat_order % len(speakers)
    return speakers[(start + sentence_index) % len(speakers)]
